[WIP] Add ability to set solid name when exporting STL

DeepHorizons opened the pull request:
Attempt at #4977

Still need to hook this up to something that will determine the solid name. As suggested in the issue a special variable could be set, `$solid_name`, but I'm not sure where to look to start modifying that.

DeepHorizons committed:
Add ability to set solid name when exporting STL

This still needs to hook up to something that can read the file and
determine the proper name

t-paul commented:
As mentioned many times before, using `$` variables will get my veto, the case @gsohler mentions is just one of those. I think the way to solve that topic is by adding a declarative way of expressing this information. Some ideas collected in https://github.com/openscad/openscad/wiki/Meta-Data-Use-Cases

jordanbrown0 commented:
I'm not a  fan of using $ variables here either, but mostly I consider that a design detail.

I think the more obvious structure is to follow the pattern established by `color()`, that a part designation is just a node in the tree:
```
part("a") { ... }
part("b") { ... }
```
Even that simple case presents a few problems:  you have to convince the processing to *not* union them.  (But what should happen if they overlap?  The strongly-related problem of assigning color for multi-color prints really wants an answer to that question.)

More complex cases lead quickly to constructs where I don't know what they mean:
```
part("a") { ... part("a1") ... }
```
```
difference() {
    part("a") { ... }
    part("b") { ... }
}
```
```
part("a") { ... }
part("a") { ... }
```

Note that those cases are questions independent of whether the subtrees are designated using $ variables or modules or some new kind of annotation.

What I would *like* to do is to find a solution for the "one part" problem, where we just want to control the internal name of an STL or the base file name of any export, in a way that will extend naturally to the multi-part case.

DeepHorizons commented:
It sounds like a section of the `part(...)` syntax should be implemented then. Something that is forward compatible with what is wanted. I was thinking for a single part:

```
part("my_part");
```

This would set the solid name for the file, not worrying about the multi-part section. This should be able to be expanded to handle the `part("my_part") {...}` syntax. The question becomes how do we search for this during the export? Im not sure if we can search for it during the export phase as it seems it only has access to `Geometry` and the like which does not seem to expose this kind of information. This means we will probably have to poke the hole in `exportFile(...)` in `export.cc` and all the variants in the calls. Im still wrapping my head around `do_export(...)` in `openscad.cc` to get an idea of what work would need to be done to get this in and any pointers would be great. Am I going to have to modify the parser? How do I make a new node?

t-paul commented:
There is no "setting" of values on geometry, there is no program running as in for example python. The `part()` would be referenced in the AST, the syntax tree representing the description (source code) of the geometry.

And that's why I'm still not convinced nesting tree elements is a good idea. It certainly will work for geometry, but it can't distinguish between module declaration and module instantiation, covering only the latter.

The problem with `part` is that it will take the name out of the global namespace, extending this for all sorts of other information will end up with a horrible mess of names taken in the global module namespace. This could be reduced by using one module name and giving it maybe key/value or something. Thing is pretty much all languages, including new designed ones have some sort of annotations exactly for that.

jordanbrown0 commented:
To be clear:  `part(...)` would turn into an AST node and then, when executed, would turn into a CSG node.  (Just like `color(...)` and most other module invocations.)

Yes, it eats into global namespace.  Ideally, a bunch of related concepts would be covered by that one module:  color, material, separate sub-models to be printed separately, maybe slicing hints, et cetera, using named parameters.  But it **doesn't** require new syntax.  Any scheme would require either new syntax or eating into global namespace.

`part("a");` does not seem like a viable stepping-stone, since in a later version with multi-part support it would need to yield a part that has no content.  But maybe if you want part naming, and we're in that interim one-part state, you must have exactly one `part()` invocation and it must be at the top level, and the entire model must be a child.

DeepHorizons committed:
Search for a part module at the top level and set that as the name

DeepHorizons committed:
Moved render and export logic into its own function

DeepHorizons committed:
Allow parts to have children and export those as separate parts

Allows the following syntax

```
part("part1") {
	sphere(3);
}
```

Which will generate a `part1.<filename>.<extension>` (`part1.output.stl`
for example)

DeepHorizons commented:
I have a working example for `part(...) {...}` syntax. The `do_export` function had to be broken up so I could call the export functionality multiple times for multiple nodes. I do a search for `part` in the children of the root node and those are exported as their own files with the name of the part prepended to the filename. One issue with the implementation is that it only handles `part(...) {...}` at the top level, parts that are not at the top level are ignored.

The following `part.scad` code compiles into two parts, `part1.output.stl` and `part2.output.stl`
```
part("part1") {
    sphere(3);
}
part("part2") {
    cube(3);
}
```
With a command line:
```
./openscad -o output.stl part.scad
```

DeepHorizons committed:
Reorder functions, I think this will make the git diff better

t-paul commented:
While I'm still not convinced it's the best idea to use `part()` as it will not cover any of the many other requested use cases, it should be able to cover one useful case.

If this is going to include multi-file export (which would be very nice, even if it's initially for STL only!), please update the title of the PR and note that this will require some additional work as that PR will need to have at least some test cases and this scenario may not be fully covered by the test suite yet. I assume it's not a complicated thing, but still some extra topic to look at.

t-paul commented on src/openscad.cc:
```diff
@@ -523,8 +538,55 @@ int do_export(const CommandLine& cmd, const RenderVariables& render_variables, F
+        did_part_export = true;
+        Tree tree(c, fparent.string());
+        auto filename2_str = solid_name + "." + filename_str;
+        // TODO how to report multiple exports?
```
Same as multi-format exports:
```
rc |= cmdline(cmd);
```

t-paul reviewed (changes requested):
I'm not sure if "<partname>.<cmdlinename>.<extension>" is the best order of names. It feels like maybe swapping the first two would be better, so `-o model.stl` would result in `model.part1.stl`, `model.part2.stl`.

DeepHorizons commented on src/core/primitives.cc:
```diff
@@ -721,6 +721,24 @@ static std::shared_ptr<AbstractNode> builtin_polygon(const ModuleInstantiation *
+
+  Parameters parameters = Parameters::parse(std::move(arguments), inst->location(), {"name"});
+
+  const auto& solid_name = parameters["name"];
```
I originally used `name` but `name()` is already used on the AbstractNode, resulting in a compile error.

DeepHorizons committed:
Make sure the name given is a string

DeepHorizons committed:
Make the part module experimental

DeepHorizons committed:
Match the parsed parameter in the part module

DeepHorizons commented on src/gui/MainWindow.cc:
```diff
@@ -2113,6 +2113,7 @@ ExportInfo createExportInfo(FileFormat format, const QString& exportFilename, co
   exportInfo.format = format;
   exportInfo.fileName = exportFilename.toLocal8Bit().constData();
   exportInfo.displayName = exportFilename.toUtf8().toStdString();
+  exportInfo.solidName = "fromMainWindowcc";
```
This was a holdover from when I was debugging how exporting works. All my testing is in the CLI path and I have not done any testing exporting from the GUI yet. This is in the path of the GUI export and will probably be removed once I start getting to GUI exports.

DeepHorizons commented on src/io/export.h:
```diff
@@ -101,7 +102,8 @@ bool is2D(const FileFormat format);
 void export_stl(const std::shared_ptr<const Geometry>& geom, std::ostream& output,
-                bool binary = true);
+                bool binary = true,
+                std::string solidName = "OpenSCAD Model");
```
Do you have an idea of a better location? I figured with the `binary` default value being set here it would make sense to also put the `solidName` default value here as well.

DeepHorizons commented on src/io/export_stl.cc:
```diff
@@ -278,17 +278,17 @@ uint64_t append_stl(const std::shared_ptr<const Geometry>& geom, std::ostream& o
   if (binary) {
-    char header[80] = "OpenSCAD Model\n";
-    output.write(header, sizeof(header));
+    output << solidName << "\n";
```
Would we want to limit the solid name to 80 characters for both binary and ascii STL?

DeepHorizons committed:
Add warnings and checks for proper STL export

DeepHorizons committed:
Output part as <output>.<part>.<ext> rather than <part>.<output>.<ext>

DeepHorizons committed:
Report multiple export codes

DeepHorizons commented on src/openscad.cc:
```diff
@@ -523,8 +538,55 @@ int do_export(const CommandLine& cmd, const RenderVariables& render_variables, F
+}
+
+
+int render_and_export(
```
Is there anything special I need to do to make them available? Where do I tell it the filename?

DeepHorizons committed:
Add comment and update solid name to be the original string for ascii
stl

DeepHorizons commented on src/io/export_stl.cc:
```diff
@@ -278,17 +278,17 @@ uint64_t append_stl(const std::shared_ptr<const Geometry>& geom, std::ostream& o
   } else {
     setlocale(LC_NUMERIC, "C"); // Ensure radix is . (not ,) in output
-    output << "solid OpenSCAD_Model\n";
+    output << "solid " << solidName << "\n";
```
I set the limit to 80 characters, 79 + newline, for both binary and ascii. I also filter out whitespace characters.

gsohler commented:
i think for binary stl you definitely have to terminate with '\0' instead
of newline, else the newline becomes part of the solid name, which might
confuse other programs,


On Wed, Feb 21, 2024 at 4:30 AM Joshua Milas ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/io/export_stl.cc
> <https://github.com/openscad/openscad/pull/4986#discussion_r1496848585>:
>
> >      char tmp_triangle_count[4] = {0, 0, 0, 0}; // We must fill this in below.
>      output.write(tmp_triangle_count, 4);
>    } else {
>      setlocale(LC_NUMERIC, "C"); // Ensure radix is . (not ,) in output
> -    output << "solid OpenSCAD_Model\n";
> +    output << "solid " << solidName << "\n";
>
> I set the limit to 80 characters, 79 + newline, for both binary and ascii.
> I also filter out whitespace characters.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openscad/openscad/pull/4986#discussion_r1496848585>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACCO4MUEAKSYWSQHKZC5THTYUVS5JAVCNFSM6AAAAABDILXRY6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQOJSGAYTINBUGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

DeepHorizons committed:
Export 80 chars when doing binary stl

The header needs to be 80 characters long acording to the spec

DeepHorizons committed:
Remove TODOs

jordanbrown0 commented on src/core/primitives.cc:
```diff
@@ -721,6 +721,24 @@ static std::shared_ptr<AbstractNode> builtin_polygon(const ModuleInstantiation *
+        "module %1$s() needs a name defined", node->name());
+    return node;
+  }
+  node->solid_name = solid_name.toString();
```
We really need a canned argument-check function (perhaps additional arguments to `Parameters::parse()`) that checks for required parameters being present, correct types, et cetera.  [`check_arguments()` in builtin_functions.cc](https://github.com/openscad/openscad/blob/dc2a0d1c28e41016abf4f91ffbc4c6c707c4f64a/src/core/builtin_functions.cc#L79) is a good starting point.

While we're looking at such commonization, it would also be good to have a canned "this is intended to be a leaf node and should not have children" function, replacing the DIY in, e.g., [`builtin_cube()`](https://github.com/openscad/openscad/blob/dc2a0d1c28e41016abf4f91ffbc4c6c707c4f64a/src/core/primitives.cc#L153).

jordanbrown0 commented:
When I last thought about this, I believe I was thinking of a two-level namespace, where the arguments to `part()` would be objects.  Each parameter would provide metadata for one consumer.  Thus, for instance, the `stl` argument to `part()` would be an object containing STL-export-specific parameters, `slice` would contain slicing hints, et cetera.

This requires #4478 for a user-space syntax for creating objects.

DeepHorizons commented:
> > I would like to see if I can export just one STL with multiple parts,
> 
> Do you have any consumers that can read such a file?

I have been using OpenFOAM to do simulations that consume STL files. I have tested multi-region with ASCII STL by `cat`ing the files together and that worked as expected. I dont know about binary.

gsohler commented:
Hi DeepHorizons,
In my yesterday's result I was able to assign an individual  color to each single triangle and it even works for F6 rendering and for 3MF export of the colored Solid. Look here: https://imgur.com/a/Ux9griH its ready for export.
Its quite versatile already, for every PolySet I maintain a Material array and each Triangle has ah Material index  into this table. We could easily add more properties next to color like part name and all that information would also be available for your stl export. You could presort the triangles before actually exporting. What do you think ?

DeepHorizons commented:
@gsohler I would want to try it out but it sounds like it could work.

UBaer21 commented:
I just like to mention that there is a distinction  between  Objects and Parts - while the 3mf only know "objects" there are objects with in a model file (parts)  and separate object files (objects)

So it will be important to allow hierarchy to group parts to objects and have multiple objects. 
e.g. when slicer open 3mf files then parts positions are not changed as this can be different material colors of one object - or modifier/ supports etc. 
But objects will be automatically arranged or centered.  Also you can print selective objects  but not parts.
